packages/remark-images-download/src/index.js

```diff
@@ -0,0 +1,43 @@
+const visit = require('unist-util-visit')
+const fs = require('fs')
+const path = require('path')
+const request = require('request')
+const shortid = require('shortid')
+const url = require('url')
+
+function plugin (opts) {
+  return (tree) => {
+    if (opts && opts.downloadImage) {
+      // images are downloaded in destinationPath
+      let destinationDir = opts.downloadDestination ? opts.downloadDestination : './'
+      if (destinationDir.slice(-1) !== '/') destinationDir += '/'
+      const destinationPath = `${destinationDir}${shortid.generate()}`
+      if (!fs.existsSync(destinationPath)) {
+        fs.mkdirSync(destinationPath)
+      }
+
+      visit(tree, 'image', function (node) {
+        const parserdUri = url.parse(node.url)
+        if (parserdUri.hostname) {
+          const extension = path.extname(parserdUri.pathname)
+          const basename = `${shortid.generate()}.${extension}`
+          const destination = `${destinationPath}/${basename}`
+          downloadImage(node.url, destination, opts.maxlength)
```

Comment:
This seems unsafe because `downloadImage` is asynchronous and we're not waiting for it to complete.

Remark plugins can be async though, returning a promise. I can help with that if you need it: https://github.com/unifiedjs/unified#function-transformernode-file-next